app/services/indexer.rb

```diff
@@ -0,0 +1,92 @@
+class Indexer
+  def initialize(args)
+    @batch_size = args[:batch_size]
+    @date = args[:date]
+    @new_index_name = args[:new_index_name]
+    @client = args[:client]
+    @logger = args[:logger]
+  end
+
+  def run
+    number_datasets_processed = 0
+
+    create_new_index
+
+    Dataset.published.find_in_batches(batch_size: batch_size) do |datasets|
+      logger.info "Batching #{datasets.length} datasets"
+      bulk_index(datasets)
+      number_datasets_processed += batch_size
+    end
+
+    logger.info "Datasets indexed to #{new_index_name}"
+  end
+
+  private
+
+  attr_reader :date, :new_index_name, :batch_size, :client, :logger
+
+  def create_new_index
+    client.indices.create(
+      index: new_index_name,
+      body: { mappings: index_mapping }
+    )
+  end
+
+  def bulk_index(datasets)
+    client.bulk(
+      index: new_index_name,
+      type: ::Dataset.__elasticsearch__.document_type,
+      body: prepare_records(datasets)
+    )
+  rescue => e
+    msg = "There was an error indexing datasets:\n#{e.message}"
+    logger.error msg
+    Raven.capture msg
+  end
+
+  def index_mapping
+    {
+      dataset: {
+        properties: {
+          name: {
+            type: 'string',
+            index: 'not_analyzed'
+          },
+          uuid: {
+            type: 'string',
+            index: 'not_analyzed'
+          },
+          location1: {
+            type: 'string',
+            fields: {
+              raw: {
```

Comment:
why is this raw? I'm not sure we need to be able to do a keyword search on location. See https://github.com/datagovuk/find_data_beta/pull/313

Reply:
@hannako This has been here for a while. It is allowing an exact match on location, which is required for the location filter to work. (specifically, populating locations => if I type 'E' in the location filter field then 'England', 'England, Wales' etc should appear below).  If the requirements for the locations filter change then I imagine this mapping will require updating